### PR TITLE
Documentation/asab storage

### PR DESCRIPTION
**Pull request by mejroslav**
# Documentation for ASAB storage module

This is a first attempt to create a documentation for `asab.storage`. (There is a lot of work to be done...)

- [x] Explain the usage of `StorageService` in a way that is very pleasant for the user.
- [x] Make docstrings in `asab.storage` more descriptive.
- [x] Rename the examples so that there are 'in one place' and perhaps make them more descriptive?
- [x] Add example on encryption and decryption.

## The structure of chapters

- Configuration: a description how to configure storage
- Manipulation with databases: example usage of the common operations (upsertor, set, get, delete)
- Storing data in memory: 
- Storing data in MongoDB: examples of methods specific for MongoDB
- Storing data in ElasticSearch: **not finished yet**
-  Encryption and decryption: a description of how to encrypt and decrypt data
- Reference: a list of abstract methods and the specific methods for every storage type

Screenshot 2023-03-23 at 20-16-00 Storage — ASAB documentation

### CODE REVIEW

File: doc/asab/storage.rst [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        }
+    )
+
+The options for the storage type are:
+
+- `inmemory`: Collects data directly in memory
+- `mongodb`: Collects data using MongoDB database. Depends on `pymongo <https://pymongo.readthedocs.io/en/stable/>`_ and `motor <https://motor.readthedocs.io/en/stable/api-asyncio/asyncio_motor_collection.html>`_ libraries.
+- `elasticsearch`: Collects data using Elastic Search database. Depends on `aiohttp <https://docs.aiohttp.org/en/latest/>`_ library.
+
+Although these three databases are different, accessing the database and manipulation with collections is done by using the same methods.
```
Comment (eliska-n) [inline doc/asab/storage.rst:39]: I think this si called strategy pattern. Maybe I am wrong... It is not important for the documentation, you can try to google it :) Architecture patterns bring a lot of fun also.

Storage Service provides a unified interface for accessing and manipulating collections across multiple database technologies.

File: doc/asab/storage.rst [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+For accessing the storage, simply `add asab.storage.Module`` when initializing and register the service.
```
Comment (eliska-n) [inline doc/asab/storage.rst:41]: It is subjective what is "simple". But it is ok here, we don't need to be 100% official.

File: doc/asab/storage.rst [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            storage = self.get_service("asab.StorageService")
+
+
+Manipulation with databases
+---------------------------
+
+Upsertor
+~~~~~~~~
+
+Upsertors are used for manipulations with databases. Upsertor is an object that works like a pointer to the specified database and optionally to object id.
```
Comment (eliska-n) [inline doc/asab/storage.rst:61]: The whole Storage Service is used to manipulate with database. First sentence says nothing new.

image
I know it is not 100% true for this implementation, but it is one clear sentence on what "upsert" mean

File: doc/asab/storage.rst [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+.. code:: python
+
+    u = storage.upsertor("test-collection")
+
+The method :func:`upsertor()` create an upsertor object associated with the specified collection. It takes `collection` as an argument and can have two parameters `obj_id` and `version`, which are used for getting an existing object by its ID and version.
```
Comment (eliska-n) [inline doc/asab/storage.rst:67]: The `upsertor()` method creates... *in English* (use Grammarly at least - it is available as VSCode plugin, pro is available at TL)
Metoda `upsertor()` vytváří... *v češtině*

File: doc/asab/storage.rst [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+.. code:: python
+
+    u.set("key", "value")
+
+To execute these procedures, simply run the :func:`execute()` coroutine method, which commits the upsertor data to the storage and returns the ID of the object. Since it is a coroutine, it must be awaited.
+
+.. code:: python
+
+    object_id = await u.execute()
```
Comment (eliska-n) [inline doc/asab/storage.rst:82]: reference doccs for Upsertor that would clearly show it is coroutine is missing...?

File: doc/asab/storage.rst [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+Getting a single object
+~~~~~~~~~~~~~~~~~~~~~~~
+
+For getting a single object, use :func:`get()` coroutine method that takes two arguments `collection` and `obj_id` and finds an object by its ID in collection.
+
+.. code:: python
+
+    obj = await storage.get(collection="test-collection", obj_id=object_id)
+
+When the requested object is not found in the collection, the method raises ``KeyError``. Remember to handle this exception properly when using databases in your services and prevent them from crashing!
```
Comment (eliska-n) [inline doc/asab/storage.rst:99]: That's so sweet :slightly_smiling_face:

**Review by eliska-n — CHANGES_REQUESTED** [review]

File: doc/asab/storage.rst [hunk 7]
```diff
@@ -0,0 +1,309 @@
+.. currentmodule:: asab.storage
+
+Storage
+=======
+
+ASAB provides a service for storing data. Data can be stored in memory or in dedicated document database. For now, `MongoDB <https://www.mongodb.com/>`_ and `Elastic Search <https://www.elastic.co/>`_ databases are supported.
```
[the remark this answers is not in the document]
Comment (ateska) [reply doc/asab/storage.rst:6]: ;-)

File: doc/asab/storage.rst [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+    cursor = collection.find({})
+    while await cursor.fetch_next:
+        data = cursor.next_object()
+        pprint.pprint(data)
+
+The full list of methods suitable for this object is described in the `official documentation <https://motor.readthedocs.io/en/stable/api-asyncio/asyncio_motor_collection.html>`_.
+
+
+
+Storing data in Elastic Search
```
Comment (ateska) [inline doc/asab/storage.rst:195]: "ElasticSearch"

**Review by ateska — CHANGES_REQUESTED** [review]

**Comment by mejroslav** [comment]
I think I've reached the point where I'm satisfied with what I've done. The user should now have a good understanding of how to interact with databases, how to encrypt and decrypt data, has access to the reference, where (s)he can look at the provided methods, and he can easily find the examples that are descriptive enough (from my point of view).

**Review by ateska — APPROVED** [review]

**Pull request merged**